### ISSUE TITLE
`pageEl.scrollIntoView` does not work properly on iOS

### ISSUE THREAD

**Issue by shankari** [enketo/enketo-core#626]
In a previous issue (#331), it looks like you added `pageEl.scrollIntoView` to ensure that the form scrolls to the top on page flip. However, it looks like this is broken on iOS. In particular:
- in a regular browser (e.g. Safari), the header is hidden. See attached video comparing the way the same form is loaded on iOS and android
[scrollIntoViewAndroidvsiOS.mp4.gz](https://github.com/enketo/enketo-core/files/3180427/scrollIntoViewAndroidvsiOS.mp4.gz)
- that may be bearable, but if embedded into a more complex single screen app, the additional scroll breaks the entire page because it offsets the body (see https://github.com/e-mission/e-mission-phone/pull/563#issuecomment-492359427) and previous.

The commit for #331 added `scrollIntoView` in three locations - commenting just one of them out - the one with `pageEl` seems to fix everything.
https://github.com/e-mission/e-mission-phone/pull/563#issuecomment-492403123

Since I am manually editing the generated bundle anyway (https://github.com/enketo/enketo-core/issues/623), I commented out that line, so I am not blocked on this. But I thought it might be helpful to file an issue in case you want a longer-term fix.

I am not sure what would be a great fix though.

**Comment by shankari** [comment]
I can reproduce on desktop Safari in Responsive Design Mode as well (see attached).
[desktop_safari_broken_too.mp4.gz](https://github.com/enketo/enketo-core/files/3180453/desktop_safari_broken_too.mp4.gz)

**Comment by shankari** [comment]
interestingly, when I try on android
- with the `pageEl.scrollIntoView();` enabled, there is a small scroll, but the bottom of the page does not move, and some of the header is still visible
- with `pageEl.scrollIntoView();` disabled, there is no scroll and the header is completely visible 

So unlike when the form is displayed directly, it looks like there is some impact of the `pageEl.scrollIntoView();` when it is embedded in a view with scrolling, but it is not as severe on android as on iOS.

**Comment by shankari** [comment]
Fortunately the test case from #331 is still around and it shows the same behavior with the hidden headers.

| Page 2         | Page 3      |
|-------------|-----------|
|<img width="319" alt="Screen Shot 2019-05-14 at 6 08 59 PM" src="https://user-images.githubusercontent.com/2423263/57741909-b0bcf180-7673-11e9-999b-9fc1e909b217.png"> | <img width="316" alt="Screen Shot 2019-05-14 at 6 09 39 PM" src="https://user-images.githubusercontent.com/2423263/57741910-b0bcf180-7673-11e9-9b4c-6b2db51314b3.png"> |

**Comment by shankari** [comment]
According to the compatibility matrix, safari doesn't have support for `scrollIntoViewOptions` - I wonder if that is why the behavior is different

**Comment by shankari** [comment]
ok so I have a potential fix, but it only works for mobile browsers. Which, honestly, is probably good enough because this is only really visible on mobile browsers. I can submit a PR if I know how you distinguish between mobile and desktop browsers.

From the same mozilla page, `scrollIntoViewIfNeeded`, which is not standard and is not on a standards track, *is* supported on webkit-based browsers including android webview and safari on iOS
https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoViewIfNeeded

Switching to `pageEl.scrollIntoViewIfNeeded();` causes mobile safari to work, even for the test case in #331 and without hiding the header (see attached video). It may be worthwhile to check this in as a workaround for mobile browsers, or maybe even just for mobile safari, until it supports `scrollIntoViewOptions`. Thoughts?

**Comment by shankari** [comment]
actually, I take that back, even with `pageEl.scrollIntoViewIfNeeded();`, the header is hidden, and the UX is offset when embedded. The only thing that doesn't cause the offset is for that line to be completely commented out.

**Comment by shankari** [comment]
Here's a stackoverflow solution for the same problem. Not sure you want to implement something that complicated.
https://stackoverflow.com/a/55572009

**Comment by shankari** [comment]
Here's additional confirmation that `scrollIntoView()` with options does not work on mobile safari
https://stackoverflow.com/a/50453912

> Only Chrome and Opera fully support block and inline, not even the smoothscroll polyfill does. It's the solution we want, but sadly not the one that works. – dube Aug 27 '18 at 14:51
> I can confirm this does not work on mobile safari as of today. window.scrollBy works. – Fabian von Ellerts Jan 7 at 16:51

**Comment by MartijnR** [comment]
Thanks for reporting this and your research! We'll read through it. I just wanted to let you know it's not being ignored but since it's iOS-specific it gets a little lower priority.

**Comment by shankari** [comment]
pedantic nitpick: it is safari-specific, not iOS-specific. I can see the same behavior on Safari for OSX as well (https://github.com/enketo/enketo-core/issues/626#issuecomment-492462456). It _is_ less likely to be visible because the larger screens mean less need for scrolling, so no quibbles with the lower priority :)

**Comment by MartijnR** [comment]
Hi @shankari, 

I don't seem to be able to reproduce the original #331 issue. Was it not scrolling at all?

If not, I wonder if we could define a custom scroll function that first calls scrollIntoView and subsequently scrollTo.

If it was scrolling but not correctly could we use the original scrollTo but scroll to the top of the form header element instead of 0,0?

Another thought: Or could we simply call scrollIntoView on a different Element (form header)?

A PR would be very welcome (probably scrolling to the first error, doesn't need to be fixed/changed).

**Comment by shankari** [comment]
@MartijnR I can reproduce #331 using Safari on OSX.

Steps to reproduce:
- Start an `enketo-core` development server (running at localhost:8085)
- Comment out the line `pageEl.scrollIntoViewIfNeeded();` around line number 302 in `src/js/page.js`
- Modify the test URL from #331 to match the new format - it is now http://localhost:8005/?xform=https://raw.githubusercontent.com/alxndrsn/enketo-client-side-transformer/master/xforms/pregnancy.xml
- Open a new tab in Safari for OSX and "Enter responsive design mode", or open Safari in the iOS emulator
- Open the URL from step 2 and navigate through the form. The first and second pages are fine, but when you go from the second to the third page, you stay at the bottom of the screen
- Restore the line
- Restart the survey - the first page is fine, but the header is hidden on the second page.

See attached videos

| With `scrollIntoView` | Without `scrollIntoView` |
| --------------------- | ------------------------- |
| page 1 -> 2 transition is broken because page moves too far up and hides the header | page 2 -> 3 transition is broken because page stays at the same scroll location  | 
| [with_scrollIntoView.mp4.gz](https://github.com/enketo/enketo-core/files/3199655/with_scrollIntoView.mp4.gz) | [without_scrollIntoView.mp4.gz](https://github.com/enketo/enketo-core/files/3199656/without_scrollIntoView.mp4.gz) |

**Comment by shankari** [comment]
Trying some of the fixes from stackoverflow to see if any of them work.
@MartijnR would you be open to a browser specific check, similar to
`if (/iPad|iPhone|iPod/.test(navigator.userAgent) {`

I am not blocked by this, so I don't have a lot of time to spend on looking at this right now, but will spend a short amount of time to see if there is a fix.

**Comment by shankari** [comment]
ok so I put a breakpoint at the line `( header || pageEl ).scrollIntoView();` ensured that header was non null

```
<header class="form-header clearfix">
...
</header>
```

and then manually executed 

```
> header.scrollIntoView();
< undefined
```

The app was again displaced. I strongly believe that `scrollIntoView` is essentially the problem

**Comment by shankari** [comment]
Yup, `scrollIntoView` is the issue. If I use the suggested implementation in https://stackoverflow.com/a/55572009 the form header is hidden, but the page is not displaced. Please see attached video.
The diff was fairly straightforward, and can be even simpler if we don't need smooth scrolling.
Note that the change was in the manually generated enketo bundle.

```
diff --git a/www/manual_lib/enketo/enketo-bundle.js b/www/manual_lib/enketo/enketo-bundle.js
index d28177d4..58568a3c 100644
--- a/www/manual_lib/enketo/enketo-bundle.js
+++ b/www/manual_lib/enketo/enketo-bundle.js
@@ -23762,7 +23762,24 @@ var enketocore = (function () {
                    .trigger( 'fakefocus' );

                // pageEl.scrollIntoView();
+            const scrollLayer = this.form.view.html.closest( '.overflow-scroll' );
+            this._scrollToElement(scrollLayer, pageEl.offsetTop, 200);
            },
+        _scrollToElement(scrollLayer, destination, duration) {
+            if (duration <= 0) {
+                return;
+            }
+            const difference = destination - scrollLayer.scrollTop;
+            const perTick = (difference / duration) * 10;
+
+            setTimeout(() => {
+                scrollLayer.scrollTop = scrollLayer.scrollTop + perTick;
+                if (scrollLayer.scrollTop === destination) {
+                    return;
+                }
+                this._scrollToElement(scrollLayer, destination, duration - 10);
+            }, 10);
+        },
            _toggleButtons( index ) {
                const i = index || this._getCurrentIndex(),
                    next = this._getNext( i ),
```

With change from #628 (note displacement around 00:14, right after executing `scrollIntoView`):
[with_scrollIntoView_displacement.mp4.gz](https://github.com/enketo/enketo-core/files/3210218/with_scrollIntoView_displacement.mp4.gz)

With change from stackoverflow (https://stackoverflow.com/a/55572009, note no displacement):
[with_stackoverflow_solution_no_displacement.mp4.gz](https://github.com/enketo/enketo-core/files/3210222/with_stackoverflow_solution_no_displacement.mp4.gz)

**Comment by shankari** [comment]
Note that the change was in the manually generated enketo bundle since that is what I use in the embedded app. If we go with this alternative, I am not sure where to add the new function in the enketo code. The obvious place would be in `src/js/app/js`.

However, we probably want to also use it to replace other instances of `scrollIntoView`, notably `target.scrollIntoView` in `src/js//form.js` What is the best place to put this function so that it can be reused across multiple modules?

**Comment by MartijnR** [comment]
I am confused. We're not using the unsupported options for scrollIntoView. We're not using any parameter. So I don't understand why we need some ugly iOS hack.

**Comment by shankari** [comment]
sorry, let me clarify. The issue appears to only occur when all these conditions are met:
- the survey is being embedded into a more complex application - e.g. this is not a survey-only app
- the application is a single-page app - e.g. angular/ionic - that uses tabs to navigate between screens
- the mobile OS is iOS

For non-embedded surveys (e.g. using mobile safari directly), the page still appears to be offset (https://github.com/enketo/enketo-core/pull/628#issuecomment-495008085), but since it is not part of some larger whole, it is not obvious and the user can just scroll down again 

In the single-page mobile app case, when `scrollIntoView()` is called on pretty much any element, the entire HTML body is moved up slightly, leaving an extra gap at the bottom of the screen (see the extent of the HTML body in  https://github.com/e-mission/e-mission-phone/pull/563#issuecomment-492359427) and the video linked above (https://github.com/enketo/enketo-core/files/3210218/with_scrollIntoView_displacement.mp4.gz). This not only looks ugly, it is a persistent change that the user cannot fix by scrolling down, since the entire page has moved.

Apparently, there is a known issue that `scrollIntoView()` scrolls just a bit too far
https://stackoverflow.com/questions/24665602/scrollintoview-scrolls-just-too-far

There are many potential solutions listed on stackoverflow
- One seems to be to use `scrollIntoViewOptions` (https://stackoverflow.com/a/49710815) but that is not supported by Safari on iOS
- The other is to essentially use other positioning techniques such as:
   - using `window.scrollTo` to move down by 10 pixels (https://stackoverflow.com/a/42056880)
   - using `scrollTop` (https://stackoverflow.com/a/24665651)
   - the one that I implemented ....

As I said, the one that I implemented is more complex than it needs to be - if it is simplified, I think that it essentially boils down to using `scrollTop`. I didn't experiment with the `window.scrollTo` but if implemented as a reusable function, it is unlikely to be a lot less complex.

If this feels like too much of a corner case to support in the main branch, I am happy to make the change only to my fork. But if you would like me to submit a PR, I am happy to do so as well :)

**Comment by shankari** [comment]
>  The diff was fairly straightforward, and can be even simpler if we don't need smooth scrolling.

Simplifying to remove smooth scrolling and to simplify the interface, I get:

```
diff --git a/www/manual_lib/enketo/enketo-bundle.js b/www/manual_lib/enketo/enketo-bundle.js
index d28177d4..ef3f299a 100644
--- a/www/manual_lib/enketo/enketo-bundle.js
+++ b/www/manual_lib/enketo/enketo-bundle.js
@@ -23762,7 +23762,12 @@ var enketocore = (function () {
                    .trigger( 'fakefocus' );

                // pageEl.scrollIntoView();
+            this._scrollToElement(pageEl);
            },
+        _scrollToElement(pageEl) {
+            const scrollLayer = this.form.view.html.closest( '.overflow-scroll' );
+            scrollLayer.scrollTop = pageEl.offsetTop;
+        },
            _toggleButtons( index ) {
                const i = index || this._getCurrentIndex(),
                    next = this._getNext( i ),
```

This does not have be an iOS specific hack. It works on both android and iOS

**Comment by MartijnR** [comment]
Thank you @shankari. I have an alternative suggestion. In the code you'll find a call to `pageEl.dispatchEvent( events.PageFlip() );`. This is fired after scrolling. I think this gives the embedders the tool they need to fix scrolling upon page flip.

I'm guessing the other calls to scrollIntoView() aren't actually a problem. 

Your fix won't be acceptable because of the `.overflow-scroll` lookup.

**Comment by shankari** [comment]
sorry for the repeated updates, but I'm trying various options to figure out exactly what is going on.
Annnnd, I now do.

Since the `overflow-scroll` hardcoding is sub-optimal, I tried to use the body instead, and it broke in exactly the same way (see video).

scroll the `body` element: everything moves up 
[body_scroll_moves_up.mp4.gz](https://github.com/enketo/enketo-core/files/3214089/body_scroll_moves_up.mp4.gz)

scroll the element with class `overflow-scroll`: content moves up, tabs are unchanged
[overflow_scroll_tab_stays_in_place.mp4.gz](https://github.com/enketo/enketo-core/files/3214090/overflow_scroll_tab_stays_in_place.mp4.gz)

So basically, it looks like the implementation of `scrollIntoView` on iOS scrolls the entire body contents. This is kind of OK as long as the only thing on the page is the survey. But if there are other things in the page, like a top or bottom set of tabs, then it moves up too and everything looks ugly.

The challenge is that in single page apps, the entire page is not supposed to scroll. The headers and footers are fixed, and there is another `div` (defined using the `ion-content` directive in ionic, but different in other frameworks) that is supposed to actually scroll.

So when we want to scroll the `pageEl`, we need to scroll in the `ion-content` container, not the page. So I don't think that even 

> In the code you'll find a call to pageEl.dispatchEvent( events.PageFlip() );. This is fired after scrolling. I think this gives the embedders the tool they need to fix scrolling upon page flip.

is likely to work because you would still need to know which is the parent scrolling container.

@MartijnR as I said, I am not blocked by this, and even if I fix it, I can just fix it in my copy.

But as a general fix, would you be open to a config option that would allow users to define the scroll layer? If it is not defined, it would revert to `body`. If it is defined, we would look up the scroll layer (`this.form.view.html.closest( '.overflow-scroll' );`). Then we would scroll in the scroll layer. So the default functionality will remain the same, but embedders can easily customize as long as they can add an id or a class to their scroll container.

I will not have time to contribute anything more complex at this time.

**Comment by shankari** [comment]
Just to clarify what I'm looking at

| Before the displacement | After the displacement |
| ------------------------- | ----------------------  |
| ![Simulator Screen Shot - iPhone 7 - 2019-05-23 at 13 51 29](https://user-images.githubusercontent.com/2423263/58285832-6407a380-7d62-11e9-8b49-7a7422101110.png) | ![Simulator Screen Shot - iPhone 7 - 2019-05-23 at 13 51 44](https://user-images.githubusercontent.com/2423263/58285831-6407a380-7d62-11e9-949a-e388aa713375.png) |

**Comment by shankari** [comment]
> I'm guessing the other calls to scrollIntoView() aren't actually a problem.

There are two other calls to `scrollIntoView()` in `form.js`. I had to comment out the             
`// target.scrollIntoView(); ` in `Form.prototype.goToTarget` as well otherwise I see the same behavior on validation error (see attached video).

[validation_failure_causes_displacement.mp4.gz](https://github.com/enketo/enketo-core/files/3214979/validation_failure_causes_displacement.mp4.gz)

**Comment by shankari** [comment]
One of the users of the e-mission platform solved this in a much easier way. They just changed the next button to have a `href` of the top of the page instead of `#`
https://github.com/e-mission/e-mission-phone/pull/572/files#diff-9705ca960b3ebc7505076e975fc47dc7

**Comment by shankari** [comment]
@MartijnR I think there are three options at this point.
1. I can submit a PR that has the `scrollIntoView()` -> `scrollElement()` fix with an option to specify the scroll layer. default for option is `body`
1. I can submit a PR to change `build/index.html` to change the "next" button's href
1. I can make the changes to my own copy and other users encountering these issues can apply any of the workarounds to their own copies.

Please let me know what you think!
